libfwupdplugin/fu-bluez-device.c

```diff
@@ -350,18 +409,140 @@ fu_bluez_device_read (FuBluezDevice *self, const gchar *uuid, GError **error)
  * Since: 1.5.7
  **/
 gchar *
-fu_bluez_device_read_string (FuBluezDevice *self, const gchar *uuid, GError **error)
+fu_bluez_device_read_string (FuBluezDevice *self, const gchar *uuid_str, GError **error)
 {
-	GByteArray *buf = fu_bluez_device_read (self, uuid, error);
+	GByteArray *buf = fu_bluez_device_read (self, uuid_str, error);
 	if (buf == NULL)
 		return NULL;
 	return (gchar *) g_byte_array_free (buf, FALSE);
 }
 
+/**
+ * fu_bluez_device_connect_signal:
+ * @self: A #FuBluezDevice
+ * @uuid_str: The UUID, e.g. `00cde35c-7062-11eb-9439-0242ac130002`
+ * @signal_name: The signal to connect, e.g. `g-properties-changed`
+ * @callback: The handler for this signal
+ * @error: A #GError, or %NULL
+ *
+ * Connects to a signal of a DBus object identified by a UUID and
+ * registers a callback to handle it.
+ * If successful, it saves the signal id in the uuid object.
+ *
+ * Returns: %TRUE if the signal connection was successful.
+ *
+ * Since: 1.5.7
+ **/
+gboolean
+fu_bluez_device_connect_signal (FuBluezDevice *self,
+				const gchar *uuid_str,
+				const gchar *signal_name,
+				gpointer callback,
```

Comment:
I was going to say you also need `gpointer callback_userdata` here so you can pass a pointer to `@callback` but I'm wondering if just emitting a signal from the `FuBluezDevice` itself might be more useful to the client. e.g. `g_signal_emit (self, "signal", name)`

Reply:
I agree, it's one less thing that every plugin needs to do too then.

Reply:
No, I think we'd just have `fu_bluez_device_connect_signal (FuBluezDevice *self, const gchar *uuid_str)` which would automatically call `fu_bluez_device_start_notify (self, uuid_str)` -- the signal would be sent out as a standard `g_signal_emit()` rather than by calling a callback.